### PR TITLE
core/either: Remove `EitherName` in favor of `Either`

### PR DESCRIPTION
**Pull request by thomaseizinger**
# Description

## Links to any relevant issues

#2650

## Change checklist

- [x] I have performed a self-review of my own code
~- [ ] I have made corresponding changes to the documentation~
~- [ ] I have added tests that prove my fix is effective or that my feature works~
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: core/CHANGELOG.md [hunk 1]
```diff
@@ -5,12 +5,14 @@
...
 - Drop `Sync` requirement on `StreamMuxer` for constructing `StreamMuxerBox`. See [PR 2775].
 - Use `Pin<&mut Self>` as the receiver type for all `StreamMuxer` poll functions. See [PR 2765].
+- Remove `EitherName` in favor of `Either` from the `either` crate. See [PR XXXX].
 
 [PR 2724]: https://github.com/libp2p/rust-libp2p/pull/2724
 [PR 2762]: https://github.com/libp2p/rust-libp2p/pull/2762
 [PR 2775]: https://github.com/libp2p/rust-libp2p/pull/2775
 [PR 2776]: https://github.com/libp2p/rust-libp2p/pull/2776
 [PR 2765]: https://github.com/libp2p/rust-libp2p/pull/2765
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
Comment (thomaseizinger) [inline core/CHANGELOG.md]:
```suggestion
- Remove `EitherName` in favor of `Either` from the `either` crate. See [PR 2798].

[PR 2724]: https://github.com/libp2p/rust-libp2p/pull/2724
[PR 2762]: https://github.com/libp2p/rust-libp2p/pull/2762
[PR 2775]: https://github.com/libp2p/rust-libp2p/pull/2775
[PR 2776]: https://github.com/libp2p/rust-libp2p/pull/2776
[PR 2765]: https://github.com/libp2p/rust-libp2p/pull/2765
[PR 2798]: https://github.com/libp2p/rust-libp2p/pull/2798
```

**Comment by mxinden** [comment]
> This does not compile because the blanket impl for `ProtocolName` over anything that implements `AsRef<u8>` is conflicting with the implementation on `Either`.

I don't have a strong opinion on this. Given the complication above I am inclined to just leave it as is. Any additional benefits of this patch beyond consistency?

**Comment by thomaseizinger** [comment]
> > This does not compile because the blanket impl for `ProtocolName` over anything that implements `AsRef<u8>` is conflicting with the implementation on `Either`.
> 
> I don't have a strong opinion on this. Given the complication above I am inclined to just leave it as is. Any additional benefits of this patch beyond consistency?

I'll give it a bit more thought but I don't think so.

Instead of just adding impls that we need, I can see to refactor some code to streamline, how we generate `ProtocolName`s.

It is a bit weird that we allow so many things to be `ProtocolName`s and on the trait, we define two requirements that aren't really enforced:

- Max length of 140
- Must start with `/`

Maybe we should introduce a `ProtocolName` newtype?

**Comment by mxinden** [comment]
On the meta level, I think we could do better modelling protocol names.

Part of the problem is the conflict between multistream select and libp2p-identify on the specification level. One accepts a `[u8]`, the other expects a `String`.

> Maybe we should introduce a ProtocolName newtype?

Or a unit type for each protocol name, each implementing `ProtocolName`.

**Comment by thomaseizinger** [comment]
> On the meta level, I think we could do better modelling protocol names.
> 
> Part of the problem is the conflict between multistream select and libp2p-identify on the specification level. One accepts a `[u8]`, the other expects a `String`.

Damn. Of course it wasn't that easy to fix :)

> > Maybe we should introduce a ProtocolName newtype?
> 
> Or a unit type for each protocol name, each implementing `ProtocolName`.

Nice idea but it won't allow us to enforce invariants, right?

**Comment by mxinden** [comment]
> Nice idea but it won't allow us to enforce invariants, right?

True. Would be great to enforce these at runtime. Would be wonderful to enforce these at compile time :innocent:

**Comment by thomaseizinger** [comment]
> > Nice idea but it won't allow us to enforce invariants, right?
> 
> True. Would be great to enforce these at runtime. Would be wonderful to enforce these at compile time :innocent: 

Something for a future issue? I have some ideas :)

To progress on this issue, would it be okay to transition all/most usages to std types like `&static [u8]` and `String`?

**Comment by mxinden** [comment]
> > > Nice idea but it won't allow us to enforce invariants, right?
> > 
> > 
> > True. Would be great to enforce these at runtime. Would be wonderful to enforce these at compile time innocent
> 
> Something for a future issue? I have some ideas :)

:+1: 

> To progress on this issue, would it be okay to transition all/most usages to std types like `&static [u8]` and `String`?

Given that `Either` also implements `AsRef<[u8]>` when the two subtypes implement `AsRef<[u8]>`, is the implementation of `ProtocolName` on `Either` still needed?

https://docs.rs/either/latest/either/enum.Either.html#impl-AsRef%3C%5BTarget%5D%3E

**Comment by thomaseizinger** [comment]
> > To progress on this issue, would it be okay to transition all/most usages to std types like `&static [u8]` and `String`?
> 
> Given that `Either` also implements `AsRef<[u8]>` when the two subtypes implement `AsRef<[u8]>`, is the implementation of `ProtocolName` on `Either` still needed?
> 
> https://docs.rs/either/latest/either/enum.Either.html#impl-AsRef%3C%5BTarget%5D%3E

(Un)fortunately yes. I remember trying this and it is not working.

I think the reason is in Rust's explicit strong-typing system. Just because something implements all super traits of a trait A does not mean it implements trait A by itself unless it is actively declared as such. In Golang, this would work :)

**Comment by mxinden** [comment]
> To progress on this issue, would it be okay to transition all/most usages to std types like `&static [u8]` and `String`?

I am fine with that.

File: core/CHANGELOG.md [hunk 2]
```diff
@@ -6,13 +6,16 @@
 - Change `StreamMuxer` interface to be entirely poll-based. All functions on `StreamMuxer` now
   require a `Context` and return `Poll`. This gives callers fine-grained control over what they
   would like to make progress on. See [PR 2724] and [PR 2797].
+- Remove `EitherName` in favor of `Either` from the `either` crate. See [PR 2798].
+- Replace blanket `ProtocolName` impl with specific ones on `&'static [u8]` and `Cow<'static, [u8]>`. See [PR 2798].
```
Comment (thomaseizinger) [inline core/CHANGELOG.md:10]: We could split this out into a different PR if you want.

Comment (mxinden) [reply core/CHANGELOG.md:10]: I am fine with a single pull request only.

File: core/src/upgrade.rs [hunk 3]
```diff
@@ -126,12 +128,28 @@ pub trait ProtocolName {
...
     }
 }
 
+impl<A, B> ProtocolName for Either<A, B>
+where
+    A: ProtocolName,
+    B: ProtocolName,
+{
+    fn protocol_name(&self) -> &[u8] {
+        ::either::for_both!(self, inner => inner.protocol_name())
```
Comment (mxinden) [inline core/src/upgrade.rs:149]: Neat!

File: core/CHANGELOG.md (hunk 2 above)
Comment (mxinden) [inline core/CHANGELOG.md:10]: Why not implement `ProtocolName` for `String` and `str`?

**Review by mxinden — COMMENTED** [review]
I would actually suggest making `String` the default protocol representation. See https://github.com/libp2p/rust-libp2p/issues/2831. What do you think @thomaseizinger?

**Comment by thomaseizinger** [comment]
> I would actually suggest making `String` the default protocol representation. See #2831. What do you think @thomaseizinger?

Yeah we can move to string! Should we do that as part of this PR or make another one? It is getting a bit OT from the original goal haha.

I am happy to fix #2831 first and then come back to this!

**Comment by mxinden** [comment]
I think the following order makes most sense:

1. #2831 
2. Move `ProtocolName` to `String`
3. Do the actual change here, namely to use `Either`

That said, I don't think we have to stick with that order.

**Comment by thomaseizinger** [comment]
> I think the following order makes most sense:
> 
>     1. [misc/multistream-select: Treat protocol as `String` and not `[u8]` #2831](https://github.com/libp2p/rust-libp2p/issues/2831)
> 
>     2. Move `ProtocolName` to `String`
> 
>     3. Do the actual change here, namely to use `Either`
> 
> 
> That said, I don't think we have to stick with that order.

I agree, leave it with me 😊

**Comment by thomaseizinger** [comment]
Closing in favor of #2966.
